governance/proposals/007-uniswap-v3-migration.js

```diff
@@ -0,0 +1,229 @@
+  'function allowance(address, address) external view returns (uint256)',
+]
+
+const migratorABI = [
+  {
+    inputs: [
+      {
+        internalType: 'address',
+        name: 'token0',
+        type: 'address',
+      },
+      {
+        internalType: 'address',
+        name: 'token1',
+        type: 'address',
+      },
+      {
+        internalType: 'uint24',
+        name: 'fee',
+        type: 'uint24',
+      },
+      {
+        internalType: 'uint160',
+        name: 'sqrtPriceX96',
+        type: 'uint160',
+      },
+    ],
+    name: 'createAndInitializePoolIfNecessary',
+    outputs: [
+      {
+        internalType: 'address',
+        name: 'pool',
+        type: 'address',
+      },
+    ],
+    stateMutability: 'payable',
+    type: 'function',
+  },
+  {
+    inputs: [
+      {
+        components: [
+          { internalType: 'address', name: 'pair', type: 'address' },
+          {
+            internalType: 'uint256',
+            name: 'liquidityToMigrate',
+            type: 'uint256',
+          },
+          { internalType: 'uint8', name: 'percentageToMigrate', type: 'uint8' },
+          { internalType: 'address', name: 'token0', type: 'address' },
+          { internalType: 'address', name: 'token1', type: 'address' },
+          { internalType: 'uint24', name: 'fee', type: 'uint24' },
+          { internalType: 'int24', name: 'tickLower', type: 'int24' },
+          { internalType: 'int24', name: 'tickUpper', type: 'int24' },
+          { internalType: 'uint256', name: 'amount0Min', type: 'uint256' },
+          { internalType: 'uint256', name: 'amount1Min', type: 'uint256' },
+          { internalType: 'address', name: 'recipient', type: 'address' },
+          { internalType: 'uint256', name: 'deadline', type: 'uint256' },
+          { internalType: 'bool', name: 'refundAsETH', type: 'bool' },
+        ],
+        internalType: 'struct IV3Migrator.MigrateParams',
+        name: 'params',
+        type: 'tuple',
+      },
+    ],
+    name: 'migrate',
+    outputs: [],
+    stateMutability: 'nonpayable',
+    type: 'function',
+  },
+]
+
+module.exports = async () => {
+  // impersontate timelock for testing
+  const signer = await ethers.getSigner(timelockAddress)
+
+  // parse call data for function call
+  const poolV2 = await ethers.getContractAt(
+    poolV2Abi,
+    udtWethV2poolAddress,
+    signer
+  )
+
+  // pool v2 info
+  const [token0, token1, [reserve0, reserve1], totalSupply] = await Promise.all(
+    [
+      poolV2.token0(),
+      poolV2.token1(),
+      poolV2.getReserves(),
+      poolV2.totalSupply(),
+    ]
+  )
+
+  const [{ symbol: symbol0 }, { symbol: symbol1 }] = await Promise.all([
+    await getTokenInfo(token0),
+    await getTokenInfo(token1),
+  ])
+  console.log(`Pair ${symbol0}/${symbol1} - at ${poolV2.address}`)
+
+  // how many ERC20 owned by the timelock
+  const liquidity = (await poolV2.balanceOf(timelockAddress)).div(10)
```

Comment:
actually already doing the 10% here but better to use uniswap built-in feature maybe